rfcs/2022-05-17-11532-chronicle-sink.md

```diff
@@ -0,0 +1,237 @@
+
+Unstructured log events are events in plain text with no strict schema attached.
+
+```json
+{
+  "log_text": "26-Feb-2019 13:35:02.187 client 10.120.20.32#4238: query: altostrat.com IN A + (203.0.113.102)",
+  "ts_epoch_microseconds": 1551188102187000
+}
+```
+
+Events are batched and sent in a JSON object that includes the `customer_id` and
+a `log_type`:
+
+```json
+{
+  "customer_id": "c8c65bfa-5f2c-42d4-9189-64bb7b939f2c",
+  "log_type": "BIND_DNS",
+  "entries": [...]
+}
+```
+
+`log_type` has to be one of the values returned by the [logtypes endpoint](https://cloud.google.com/chronicle/docs/reference/ingestion-api#logtypes).
+
+
+If `event_type == "unstructured"` we will require another configuration field -
+`log_type` to be set. This field will be templatable. The field can be removed
+from the final event via `encoding.exclude_fields`.
+
+The encoding of the message (text or json) can be set via an `encoding` field.
+
+##### Schema
+
+A UDM message has the following sections. (To keep this RFC succinct, this is
+not a comprehensive list, the full list can be found
+[here](https://cloud.google.com/chronicle/docs/reference/udm-field-list#udm_event_data_model).)
+
+- *Metadata*
+  Contains metadata about the event. The fields available for metadata can be
+  found [here](https://cloud.google.com/chronicle/docs/reference/udm-field-list#metadata).
+  Mandatory fields are `event_type` and `event_timestamp`.
+
+  The value of `event_type` has repercussions for which fields are mandatory
+  in the rest of the event. See [here](https://cloud.google.com/chronicle/docs/unified-data-model/udm-usage#required_and_optional_fields_based_on_event_type)
+  for the list. For example, if `event_type == "FILE_COPY"` then `src.file`
+  becomes mandatory. **Currently it is not possible to validate a field based
+  on the value of another field within Vector schema, so this will need to
+  be left for future improvement.
+
+- *Principal*
+  The principal is the entity that originates the event. This field is mandatory
+  and can contain any of the fields defined
+  [here](https://cloud.google.com/chronicle/docs/reference/udm-field-list#noun).
+
+- *Src*
+  The source entity being acted upon. Depending on the `event_type` this field
+  is optional. Can contain any of the fields defined
+  [here](https://cloud.google.com/chronicle/docs/reference/udm-field-list#noun).
+
+- *Target*
+  The target entity being acted upon. Depending on the `event_type` this field
+  is optional. Can contain any of the fields defined
+  [here](https://cloud.google.com/chronicle/docs/reference/udm-field-list#noun).
+
+### Implementation
+
+Both event types require different batching strategies.
+
+#### UDM
+
+UDM can use a simple `BatchedHttpSink` with a `JsonArrayBuffer` to perform the
+batching.
+
+#### Unstructured log events
+
+Since log events are batched by `log_type` we will need to partition the
+batches on the value of the `log_type` field using a `PartitionHttpSink`.
```

Comment:
This is a nit, but new sinks must use the "new-style" architecture: anything `BatchSink`, `PartitionBatchSink`, `JsonArrayBuffer`, etc, etc, is deprecated.

Reply:
That's a very important piece of information. Not a nit at all!